bionic/code_references.py

```diff
@@ -0,0 +1,159 @@
+
+    def set_tos(t):
+        nonlocal tos
+        if tos is not None:
+            # Hash tos so we support reading multiple objects
+            refs.append(tos)
+        tos = t
+
+    # Our goal is to find referenced objects. The problem is that co_names
+    # does not have full qualified names in it. So if you access `foo.bar`,
+    # co_names has `foo` and `bar` in it but it doesn't tell us that the
+    # code reads `bar` of `foo`. We are going over the bytecode to resolve
+    # from which object an attribute is requested.
+    # Read more about bytecode at https://docs.python.org/3/library/dis.html
+
+    for op in dis.get_instructions(code):
+        try:
+            # Sometimes starts_line is None, in which case let's just remember the
+            # previous start_line (if any). This way when there's an exception we at
+            # least can point users somewhat near the line where the error stems from.
+            if op.starts_line is not None:
+                lineno = op.starts_line
+
+            if op.opname in ["LOAD_GLOBAL", "LOAD_NAME"]:
+                if op.argval in context.globals:
+                    set_tos(context.globals[op.argval])
+                else:
+                    set_tos(op.argval)
+            elif op.opname in ["LOAD_DEREF", "LOAD_CLOSURE"]:
+                set_tos(context.cells[op.argval])
+            elif op.opname == "IMPORT_NAME":
+                try:
+                    set_tos(importlib.import_module(op.argval))
+                except ImportError:
+                    set_tos(op.argval)
+            elif op.opname in ["LOAD_METHOD", "LOAD_ATTR", "IMPORT_FROM"]:
+                if tos is None:
+                    refs.append(op.argval)
+                elif isinstance(tos, str):
+                    tos += "." + op.argval
+                else:
+                    tos = getattr(tos, op.argval)
+            elif op.opname == "DELETE_FAST" and tos:
+                del context.varnames[op.argval]
+                tos = None
+            elif op.opname == "STORE_FAST" and tos:
+                context.varnames[op.argval] = tos
+                tos = None
+            elif op.opname == "LOAD_FAST" and op.argval in context.varnames:
+                set_tos(context.varnames[op.argval])
+            else:
+                # For all other instructions, hash the current TOS.
+                if tos is not None:
+                    refs.append(tos)
+                    tos = None
+        except Exception as e:
+            message = oneline(
+                f"""
+            Bionic found a code reference in file ${code.co_filename}
+            at line ${lineno} that it cannot hash when hashing
+            ${code.co_name}.This should be impossible and is most
+            likely a bug in Bionic. Please raise a new issue at
+            https://github.com/square/bionic/issues to let us know.
+            """
+            )
+            raise CodeVersioningError(message) from e
```

Comment:
Like we discussed a while back, I think we should give users the option to ignore bytecode of an entity. Since we don't have that yet, I haven't added anything related to it in the error message. But I can add a code comment (or a TODO) so we don't forget about it.